Windows: Add-Warning() and Add-DeprecationWarning()

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
module_utils/powershell.ps1

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
Added an Add-Warning() function to assist modules in managing warnings. This functions ensures that if there is not warnings-attribute (or if it isn't a list) the attribute is added correctly and the message is appended to the list.

Similar to what already exists for python modules.

dagwieers committed:
Windows: Add Warn() and Deprecate() mechanisms

Similar to what already exists for python modules.

jhawkesworth commented:
This is great, but just wondering if we should stick with powershell naming convention i.e. Verb-Noun so instead of Warn you'd probably have something like Write-Warning or perhaps Write-AnsibleWarning?

Personally its not something I'm particularly keen on but worth considering, because as soon as its out there its going to get used (has been already in your other PRs) - its effectively adding to the ansible module (powershell) API.  Its a tricky balance as your function names would be more familiar to python module authors.

dagwieers commented:
The whole Windows module infrastructure is going to be rewritten by @nitzmahone anyhow, but in this case I wanted to stay close to the python modules. I try to do the same for the option to Set-AnsibleParam for instance, I hate the `--failifempty` (should have been `-required`) or `-ValidateSet` (should have been `-choices`).

So I understand that the Windows/Powershell world has their own way of doing things, but I wouldn't deviate from what Ansible is doing. IMO it is going to be much easier to document this if we stick to common practices as much as possible and doesn't alienate one group from the other.

I would have preferred warn() instead, but that's less an issue because Powershell doesn't care anyhow :-)

bcoca commented:
@nitzmahone is on PTO right now, but I believe he'll agree that we keep the API as close as possible.

dagwieers committed:
Turn deprecations from list of strings, to list of dicts

Since #20884 the internal representations of deprecation messages is
changed from a list of strings to a list of dicts.

jhawkesworth commented:
Wouldn't want to hold up this going in, just wanted to flag up concern.  More important to have the api capability than to make it instantly recognisable to those familiar with powershell.

shipit

dagwieers commented:
@jhawkesworth No worries, we can wait on @nitzmahone 's return to decide. I know some of the interfaces will be changed in the future, but since we don't know whether this will hit v2.3 or later I prefer we don't focus too much on that right now and add as much functionality to make it on-par with Python modules.

It is going to be much easier to replace A with B once it is needed, then to have to revisit the code and refactor such support into it after the fact.

nitzmahone commented:
My only issue with adding new global functions to to the API surface area at this point is that anything we add will have to get dragged around for at least two releases before we could remove it (since we can't know what private modules are using). OTOH, since the new wrapper supports arbitrary module_utils inclusions (and powershell.ps1 is not strictly required anymore), anything new won't be defined in powershell.ps1 anyway, so we can pretty much consider that file "legacy" already. 

As far as naming goes, I think @jhawkesworth is on the right track- I'd vote for `Add-Warning`/`Add-DeprecationWarning` (since we're not "writing" it per-se like Write-Output/Error/etc). 

I'm OK to merge this with that name change (or feel free to suggest others)...

dagwieers committed:
Rename to Add-Warning() and Add-DeprecationWarning()

Implemented as discussed.

dagwieers commented:
@nitzmahone Thanks for the feedback. Implemented as requested.

jhawkesworth commented:
Nice.

shipit

nitzmahone commented on lib/ansible/modules/windows/win_regedit.ps1:
```diff
@@ -37,12 +37,11 @@ $result = @{
 # Fix HCCC:\ PSDrive for pre-2.3 compatibility
 if ($path -match "^HCCC:\\") {
-    $result.warnings += "Please use path: HKCC:\... instead of path: $path\n"
+    Warn $result "Please use path: HKCC:\... instead of path: $path"
```
Needs rename :)

dagwieers replied:
I didn't realize these changes were in this PR too.

nitzmahone commented on lib/ansible/modules/windows/win_command.ps1:
```diff
@@ -105,6 +103,11 @@ $util_type = Add-Type -TypeDefinition $util_def
 
 # FUTURE: extract this code to separate module_utils as Windows module API version of run_command
 
+$result = @{
```
I don't really have a problem with this placement, but this seems like an arbitrary and completely unrelated change.

dagwieers replied:
I didn't realize these changes were in this PR too.

nitzmahone commented on lib/ansible/modules/windows/win_shell.ps1:
```diff
@@ -80,6 +78,11 @@ If($removes -and -not $(Test-Path $removes)) {
 
 Add-Type -TypeDefinition $helper_def
 
+$result = @{
```
Ditto

dagwieers replied:
I didn't realize these changes were in this PR too.
